### ISSUE TITLE
User/Authentication system

### ISSUE THREAD

**Issue by matthijs110** [matthijs110/shopping-cart-extended#2]
The [`Customer`](https://github.com/matthijs110/shopping-cart/blob/master/app/Customer.php) is the actual `User`. So the it should be [`Authenticatable`](https://github.com/illuminate/auth/blob/master/Authenticatable.php). A few points that need attention: 

* The current [`customers`](https://github.com/matthijs110/shopping-cart/blob/master/database/migrations/2016_06_03_122421_setup_cart_tables.php#L60-L68) table does not contain information like a username or password.
* The [`Customer`](https://github.com/matthijs110/shopping-cart/blob/master/app/Customer.php) model should be [`Authenticatable`](https://github.com/illuminate/auth/blob/master/Authenticatable.php).
    * A few modifications to Laravel to allow the [`Customer`](https://github.com/matthijs110/shopping-cart/blob/master/app/Customer.php) model to be [`Authenticatable`](https://github.com/illuminate/auth/blob/master/Authenticatable.php).
    * (Obviously) Setup the required tables, routes etc..
* The possibility to proceed as a guest (one-time-purchase)
    * If the user changed his mind and creates an account later on (not within the same purchase), attach the existing [`address`](https://github.com/matthijs110/shopping-cart/blob/master/database/migrations/2016_06_03_122421_setup_cart_tables.php#L70-L80) row to the user. (Prevent duplicate rows)

**Comment by matthijs110** [comment]
b5e4b00 First steps made 👍

**Comment by matthijs110** [comment]
I think it will be harder than I thought to convert the [`User`](https://github.com/matthijs110/shopping-cart/blob/master/app/User.php) to a [`Customer`](https://github.com/matthijs110/shopping-cart/blob/master/app/Customer.php). 

Currently: all the customer info from a other will be saved inside the `customer` table. If that becomes the [`authenticatable`](https://github.com/illuminate/auth/blob/master/Authenticatable.php) model, where do we put the "guests" (order without a registered user? The guest information (name, address etc..) should be stored somewhere (else) :/

**Comment by March3se** [comment]
Do you think to create also the user authentication system?

**Comment by March3se** [comment]
I tried to create it but I'm having problems

**Comment by DCzajkowski** [comment]
This thread is about making the authentication system. It seems to be harder than Matthijs suspected (I also tried at some point), but neither him nor I have the time to take care of it. If you can make it and submit a PR that would be great! If not, you have to wait for either of us to find a solution or someone from the community. Cheers :)

**Comment by March3se** [comment]
We can do this together?

**Comment by DCzajkowski** [comment]
We could, but I don't have time at all. Make a PR if you have done anything and we could improve on it. Thanks for your willingness to help :)

**Comment by March3se** [comment]
This is what I was thinking. Thank you for the time you have devoted to answer me.

**Comment by March3se** [comment]
How can I join this project. I would have created something but I would like to have your opinion, obviously if you have time.

**Comment by DCzajkowski** [comment]
1. Fork this project (button in the upper-right corner)
2. Go to your forked project.
3. Add, edit, delete etc. Whatever will help.
4. Make a pull request

If your changes make sense and are PSR compliant the PR will be most likely merged and your changes will take effect.

You can edit whatever you feel is necessary and will help its growth/functionality. :)

**Comment by March3se** [comment]
Thank you very much. I try to do it now. Although it's a little late. Tomorrow school! :(

**Comment by March3se** [comment]
If you have pleasure to join you will make me very happy. :) I'd like to get your opinion. Thanks again

https://github.com/March3se/shopping-cart-extended

**Comment by abr4xas** [comment]
Hi there, what do you think about this: 

![image](https://cloud.githubusercontent.com/assets/405484/20653281/7e9718da-b4e1-11e6-8920-243dba42df0a.png)

At this moment i'm working on fake data to test this shopping cart. I think maybe tomorrow this will be ready :smile:

**Comment by matthijs110** [comment]
Looks good. But it would be nice if people could also make an order without registering: just giving their address info.

**Comment by abr4xas** [comment]
@matthijs110 yes, i was thinking the same thing... In this case, customer model shouldn´t be authenticatable...

**Comment by DCzajkowski** [comment]
So a `User` belongs to `Customer` then?

I guess the customer then will be anyone and the User just the registered user. **But.** Doesn't that lead to duplication of data? Even a pretty serious one.

**Comment by matthijs110** [comment]
Good point @DCzajkowski. We don't want duplicate data. So we need to think about a way to store the same data, but not depending on the User. 

What about giving the `Customer` table a `user_id` which can be null. If the user decides to register: create a new user and assign the `user_id` to the `Customer` row.
